push_gateway_pruner.py

```diff
@@ -0,0 +1,152 @@
+import requests
+import logging
+import base64
+import sys
+import threading
+import urllib.parse
+from typing import Generator, Dict
+from datetime import datetime
+
+import click
+from prometheus_client import parser, Metric
+from prometheus_client.samples import Sample
+
+logger = logging.getLogger()
+
+
+@click.command()
+@click.option(
+    "--prune_interval_seconds",
+    default=600,
+    type=int,
+    help="How often to start the pruning job in seconds. Default is 600.",
+)
+@click.option(
+    "--pushgateway_url",
+    default="http://localhost:9091/",
+    type=str,
+    help="URL for your pushgateway instance (without /metrics postfix). Default is http://localhost:9091/",
+)
+@click.option(
+    "--prune_threshold_seconds",
+    default=600,
+    type=int,
+    help="Number of seconds since <prune_metric_name>'s value after which to prune a sample. Default is 600.",
+)
+@click.option(
+    "--get_request_timeout",
+    default=2000,
+    type=int,
+    help="Timeout in seconds for the Get request to <pushgateway url>/metrics. Default is 2000.",
+)
+@click.option(
+    "--prune_metric_name",
+    default="push_time_seconds",
+    type=str,
+    help="The name of the metric which is used as each group's start time. Default is 'push_time_seconds'.",
+)
+def start_interval(
+    prune_interval_seconds,
+    pushgateway_url,
+    prune_threshold_seconds,
+    get_request_timeout,
+    prune_metric_name,
+):
+    threading.Timer(prune_interval_seconds, start_interval).start()
+    prune(
+        pushgateway_url, prune_threshold_seconds, get_request_timeout, prune_metric_name
+    )
+
+
+def prune(
+    pushgateway_url: str,
+    prune_threshold_seconds: int,
+    get_request_timeout: int,
+    prune_metric_name: str,
+):
+    logger.info(
+        f"Starting pruning for {pushgateway_url} with prune threshold = {prune_threshold_seconds} seconds"
+    )
+    raw_metrics = get_metrics(pushgateway_url, get_request_timeout)
+    metrics = parse_raw_metrics(raw_metrics)
+    filtered_samples = filter_metrics(
+        metrics, prune_metric_name, prune_threshold_seconds
+    )
+    delete_old_samples(filtered_samples, pushgateway_url)
+    logger.info(f"Done with push_gateway_pruner.")
+
+
+def get_metrics(pushgateway_url: str, get_request_timeout: int) -> str:
+    url = urllib.parse.urljoin(pushgateway_url, 'metrics')
+    res = requests.get(url, timeout=get_request_timeout)
+    if res.status_code != 200:
```

Comment:
![image](https://user-images.githubusercontent.com/237927/155724420-3b58d512-1562-40b3-8be0-d5fcad4e1bda.png)
https://stackoverflow.com/questions/16511337/correct-way-to-try-except-using-python-requests-module
Use something like this for properly handling requests that might fail.